### ISSUE TITLE
Emacs hangs on large outputs from tsserver

### ISSUE THREAD

**Issue by domachine** [ananthakumaran/tide#184]
Hi,

thanks for this amazing project. One thing keeps bugging me though. From time to time emacs starts freezing and a message pops up saying that the undo buffer for the `*tide-server*` buffer becomes too large. From that time on I started to manually call `buffer-disable-undo` on the `*tide-server*` buffer once the ts server started up. This stops emacs from freezing. I wondered if it's safe to disable `undo` for this buffer automatically once the mode is loaded. I'll happily hacking together a pull-request if this an option that you're fine with.

Thank you, for sacrificing your spare time!

**Comment by ananthakumaran** [comment]
go ahead and open a PR. I find no reason to enable undo in `*tide-server*` buffer.

**Comment by domachine** [comment]
Ok on further investigation, I have to admit this still occurs :-(. I just thought it was related to the undo buffer. This mainly occurs when working with the `react-apollo` library. I already found an issue on their repo (https://github.com/apollographql/react-apollo/issues/786) that I think is related. The typings seem so complex that this slows down the typescript language server in general. I'm not very familiar with the inner workings of `tide` but is there any way to handle this in a way that doesn't let emacs hang several seconds? And with hanging I mean the cursor freezes and the editor is not even responsive to `ctrl+g`.

Oh, and this is my setup by the way:

- `GNU Emacs 25.1.1 (x86_64-apple-darwin16.0.0, NS appkit-1504.00 Version 10.12 (Build 16A323)) of 2016-10-27`
- `macOS Sierra 10.12.5`

**Comment by ananthakumaran** [comment]
> doesn't let emacs hang several seconds?
This is a bad thing. Can you provide more details
1) The command which causes/triggers the problem
2) The request/response log related to that command see https://github.com/ananthakumaran/tide#debugging for details on how to capture it.
3) You can use `toggle-debug-on-quit` to get the stack trace when you quit using ctrl-g

**Comment by domachine** [comment]
This isn't actually triggered by a special command. It more or less occurs now and then when I work with a file (moving the cursor, normal typing). I've got a feeling that this happens once I pass a complex function (like `graphql` from `react-apollo`) with the cursor and that the generation of the documentation shown in the minibar is the trigger of that problem. But I currently see no deterministic way to reproduce. All I can say is that this is almost every time combined with the undo buffer message (which I now have suppressed using my configuration). Maybe it's triggered by a huge parsing overhead of the tsserver response.

I'll try to capture reasonable debug output using the techniques you described. Maybe I can also throw a repository together with typical code, libraries that I work with so that you can (hopefully) reproduce the issue.

Hope, this helps.

**Comment by nmuth** [comment]
I'm also experiencing this problem with `react-apollo`. Since emacs is too frozen to respond to `C-g`, `toggle-debug-on-quit` doesn't give a useful backtrace. I'll try to capture some debug output as well.

**Comment by josteink** [comment]
Anyone able to provide a stacktrace for where we're stuck is definitely going to help us trying to solve this.

**Comment by nmuth** [comment]
I think I found the relevant output in the `tsserver` logs. Three lines, 22 megabytes 😮. Summary:

```
Info 251  request: {"command":"quickinfo-full","seq":"2011","arguments":{"file":"/Users/nmuth/Projects/fulgid/v2.fulgid.io/pages/post/details.js","line":42,"offset":17}}
Perf 252  2011::quickinfo-full: elapsed time (in milliseconds) 17022.3773
Info 253  response: (22MB of JSON)
```

[Here's the full file on Google Drive](https://drive.google.com/open?id=0BxyeMzvR9926eVIxNlZZeEswMGc). Let me know if you can make any sense of it.

**Comment by josteink** [comment]
Oh nice. Even without looking at the code, I can make a few observations/assumptions:

* Parsing 22MB JSON is going to take more time than parsing 1KB og JSON.
* Emacs can handle huge files, but is known to have terrible performance (to the point of locking up) when *lines* get long.
* The JSON structure returned is on one line only: ![image](https://user-images.githubusercontent.com/411338/29311072-a727c302-81b0-11e7-8b6d-f46791932b8d.png)

I know from experience that a line like this is going to absolutely *kill* Emacs for any normal editing.

I'm therefore going to guess that the performance problems we're seeing here is due to Emacs struggling inside the `tsserver`-buffer with the length of the response-line, and not something inherently "bad" with `tide` itself.

If so... What can we do about it?

**Comment by nmuth** [comment]
It looks like the process filter for `tsserver` [blindly inserts received data into `*tide-server*`](https://github.com/ananthakumaran/tide/blob/master/tide.el#L430):

```
(defun tide-net-filter (process data)
  (with-current-buffer (process-buffer process)
    (goto-char (point-max))
    (insert data))
  (tide-decode-response process))
```

Does `tide` reference the output of this buffer anywhere? If not, we can probably just discard it, or write a simple log line like `recv 12345 bytes from tsserver process`.

**Comment by nmuth** [comment]
Also, reliable steps to reproduce:

- enter some JS or TS file that imports `graphql` from `react-apollo`
- use the `graphql` function:

```
const foo = graphql(gql`{ foo { bar } }`)
```
- place cursor over `graphql` invocation; `tide` should now ask `tsserver` for type information
- wait a while
- notice that the cursor stops blinking and Emacs becomes unresponsive as the response is loaded

**Comment by ananthakumaran** [comment]
> Does tide reference the output of this buffer anywhere?
check `tide-decode-response`. Basically, all the data received is appended to the end of the buffer and once enough data is available, it will be parsed.

**Comment by nmuth** [comment]
Ok. Then I think the fix is to insert line breaks that won't break the JSON. Maybe if the line is longer than 128 bytes, split it on the last `{` before writing to the buffer?

**Comment by josteink** [comment]
@nmuth In theory that sounds like a reasonable work-around, but I think it will come at the cost of some complexity. For instance... How can you make sure that the `{` character is not part of some string-data?

I'd rather try to leverage what Emacs already has of built-in functionality for this. For instance, Emacs already has `json-pretty-print` which breaks a JSON-structure into several lines:

````elisp
(defun json-pretty-print (begin end)
  "Pretty-print selected region."
  (interactive "r")
  (atomic-change-group
    (let ((json-encoding-pretty-print t)
          ;; Ensure that ordering is maintained
          (json-object-type 'alist)
          (txt (delete-and-extract-region begin end)))
      (insert (json-encode (json-read-from-string txt))))))
````

I think we should be able to use some of the functions and approaches used here directly, instead of inventing our own JSON-parser.

**Comment by domachine** [comment]
What about filtering the tsserver output through another executable that pretty-prints the json adding some kind of separator to distinguish between different records?

**Comment by josteink** [comment]
External executables complicates the initial setup and makes everything inherently less portable than pure Emacs lisp.

I'll avoid that approach unless that's the only way. 

I use both Windows and Linux and I demand my Emacs to work the same everywhere :)

**Comment by domachine** [comment]
Maybe a simple nodejs script? As far as I can see this package already
bundles the tsserver which is a nodejs program itself.

Am 15.08.2017 21:01 schrieb "Jostein Kjønigsen" <notifications@github.com>:

**Comment by josteink** [comment]
I'd rather have it all in elisp unless there's a very good reason not to.

I don't think it's going to be very hard, and I've seen no good reasons not to try that approach first. 

Duct taping various things together bash-script style is, as I've already said, something I'll avoid if possible.

**Comment by vegansk** [comment]
> How can you make sure that the { character is not part of some string-data?

@josteink ,  count " and handle { only if the count of " is even ?

**Comment by ananthakumaran** [comment]
![screen shot 2017-08-16 at 10 39 53 am](https://user-images.githubusercontent.com/149238/29348172-4d6f5dc4-826f-11e7-88a5-fc966ceabdbe.png)

Looks like it's fixed in latest version of `apollo-react`. I am not able to reproduce with `1.4.4`.

**Comment by nmuth** [comment]
@vegansk Ah, but `"` can be part of string data as well 😜

I didn't know `json.el` was included in Emacs. Seems like `json-pretty-print` is the way to go here.

@ananthakumaran I can no longer reproduce either, awesome! The long-line slowdown bug still stands, but shouldn't be nearly as big of a problem anymore.

**Comment by josteink** [comment]
> I can no longer reproduce either, awesome!

> Seems like json-pretty-print is the way to go here.

@nmuth I'm getting mixed signals here. Is this something we should fix or not?

**Comment by ananthakumaran** [comment]
> json-pretty-print is the way to go

It internally creates a buffer
```
(defun json-read-from-string (string)
  "Read the JSON object contained in STRING and return it."
  (with-temp-buffer
    (insert string)
    (goto-char (point-min))
    (json-read)))
```
> The long-line slowdown bug still stands
I don't think it's easy to fix. Even if you figured out how to parse 20Mb json in less than 100ms with emacs lisp, there are other problems like traversing data to construct output etc. I guess if nobody is going to hit the problem, there is no point in fixing it.

**Comment by nmuth** [comment]
There is still a bug here in that `tsserver` has the potential to barf out long lines that Emacs can't handle. However, users are less likely to encounter the bug because AFAIK `react-apollo` was the only widely-used project that had types this complex.

@ananthakumaran Could we hold the process output in a variable instead of saving it to the buffer?

**Comment by vegansk** [comment]
@nmuth , just ignore ```"``` that preceded by ```\```   :-)

and it'll be faster than json-pretty-print

**Comment by ananthakumaran** [comment]
> Could we hold the process output in a variable instead of saving it to the buffer?

we use json.el, acts on buffer not string. As I said earlier, even if we fix parsing, there might be other issues that might make tide/emacs hang.

**Comment by josteink** [comment]
@vegansk: Parsers are a notorious source of bugs in software everywhere. I'd rather not write any more parsers than I need to.

If you want to make a `make-json-emacs-friendly`-function and create the associated test-suite to ensure it always does the right thing... Feel free. We might use it when it's production ready.

For now it seems further work on this issue is no longer required. I propose we close it, and then we can re-open it later if this problem re-emerges.

**Comment by nmuth** [comment]
@ananthakumaran Ah, I get the problem now. Maybe it'd be easier to patch Emacs to handle long lines 😛 

@josteink That's probably the right call.

**Comment by domachine** [comment]
Amazing. I can't reproduce with `1.4.14`, too. I think it would be amazing for tide to be able to handle such large typings. But now that `react-apollo` fixed their thing, it's far from urgent 👍 . I think we should rename this issue and just keep it open so that everyone can find it, if a problem with another package occurs.

**Comment by josteink** [comment]
Sounds good. Thanks for your feedback everyone!

Closing for now. Re-open if needed.

**Comment by aaronjensen** [comment]
Hi @josteink I'm hitting this quite a bit when using the [glamorous](https://github.com/paypal/glamorous) types. Specifically, any time tide needs to display https://github.com/paypal/glamorous/blob/master/typings/css-properties.d.ts it chokes pretty hard.

**Comment by ananthakumaran** [comment]
can you check the log and see what's the size of data that is being sent between emacs & tsserver? and also add a small minimum reproducible example.

**Comment by aaronjensen** [comment]
@ananthakumaran it's about 900k.

Here's a repro: https://github.com/aaronjensen/tide-slow-repro

**Comment by josteink** [comment]
I'll try to prototype a JSON-shredder which can reformat stuff like this across multiple lines. 

It would be interesting to see if that helps resolve this issue or not.

**Comment by josteink** [comment]
For the time being, here is a very crude and primitive approach to the problem:

https://github.com/ananthakumaran/tide/tree/json-insert

It is not production-ready by any means, but here's the gist of what I did and found out:

* It "shreds" JSON into several lines by looking for property-headers and inserting line-shifts before them. It doesn't try to understand JSON in any way... And may have issues with values containing property-like declarations.
* On the sample JSON I've tested though... This rather crude approach seems to work fine.
* Doing these modifications to the JSON however causes the `tsserver`-buffer to gain more data than reported by the `Content-length` header returned by `tsserver`.
* This seems to trip up `tide-enough-response-p` and by proxy `tide-decode-response`, causing quite an amount of JSON-decoding errors on larger inputs.
* I tried forking `json-read-object` into `tide-json-read-object` to survive this condition by handling early `json-eof` without causing failure, but this was more work than I expected. And there I stopped.

Basically it has a few issues. From my experimental testing though, it does seem to reduce the amount of lock-ups, but it's still seems slow/unstable on the repro-repo provided by @aaronjensen.

If anyone wants to test it, or do further work on it, feel free. I really don't think this has to be a super-big job, but it will need some more work and definitely more testing.

**Comment by aaronjensen** [comment]
Would there be a way of detecting the length of the response and just ignoring it if it's over a certain length? Seeing the type in my minibuffer for this particular type is not very useful, it doesn't come anywhere near fitting. If this were configurable, people could adjust to their tolerance. You could display a message in the minibuffer indicating that the response was too large or something so people could debug that if they cared.

**Comment by ananthakumaran** [comment]
The length of the content is available, so it's possible to skip/delete part of that buffer without parsing it. There is one catch though, we depend on the `:request_seq` value on the response to dispatch to the correct callback.

**Comment by aaronjensen** [comment]
Hm, well, that appears to be at the beginning of the message at least. You could probably grab it w/ `string-match`

**Comment by josteink** [comment]
We already do. But way after JSON contents have been inserted into buffer, causing the Emacs hang-up.

**Comment by aaronjensen** [comment]
I see. But if the order of those two operations was reversed could prevent inserting large payloads into the buffer? 

1. Get `request_seq` using `string-match`
2. If response length > threshold, callback request w/ "too big" response
    If not, insert response into buffer as usual

I'm not familiar w/ the code though, so I'm just theorycrafting here.

**Comment by josteink** [comment]
I'm not saying it can't be done, but like you say, we would need to reverse how we process the incoming *stream* of stdout data.

And that kind of change is probably more than I can craft on the little spare time I have these days.

**Comment by philiparvidsson** [comment]
@ananthakumaran 

Why is this closed? I'm still having this problem. After a while, tide starts freezing for 30 secs up to 2 minutes. Terminating node.exe (which runs the tide server) fixes it immediately, and then tide stays fast for a while again after you restart the server.

**Comment by ananthakumaran** [comment]
Please open a new issue with details, this issue is closed because there is a config to ignore large messages from tsserver, check the readme for details about`tide-server-max-response-length`.
